Add GPU-Assisted Validation Support

karl-lunarg opened the pull request:
Add GPU-Assisted support to the core validation layer
Add VLT test for above
Add documentation
Bump glslang version to pickup required SPIRV-Tools dependencies

Please see `docs/gpu_validation.md` for a design document which may be helpful for review.

The original development branch is in karl-lunarg/Vulkan-ValidationLayers branch: karl-gpu which has all the individual commits, but is unlikely to be useful for review.  It will probably be easier to look at the design doc and the code.  The main code file is all brand new anyway.

karl-lunarg commented:
I thought of some changes that I'll probably make on Monday(12/31)

* Move the "Get more recent glslang" commit to the first commit to keep this commit set bisectable.  It is likely that the glslang known-good will get updated before merging this, so I may end up being able to drop this commit completely.

* Fix the problems noted in the Travis-CI run.

* Change the logic for binding the debug descriptor.  Instead of doing it at CmdBindDescriptorSets time, do it after the graphics pipeline is bound in CmdBindPipeline.  I should be able to look at the pipeline to see if the app is using the debug slot and if it is not, add a command to the command buffer to bind the debug set.  This gives more positive control over the situation.

karl-lunarg commented:
The first two tasks mentioned above are done. Still need to work on the third one.

jzulauf-lunarg commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,670 @@
+  * Update (write) the descriptor set with the memory info
+* Record the above objects in the per-CB state
+
+#### GpuPreCallRecordFreeCommandBuffers
```
Is a hook needed in the ResetCommandPool path? I'm not seeing one.

karl-lunarg replied:
I don't think so.  You asked about this before, I checked and didn't find anything.  I've checked again and I'm certain that the per-CB GPU tracking information "survives" a call to `ResetCommandBufferState()` because this function isn't aware of the GPU tracking info members in the CB state.

Also, the VLT code creates a CommandPool with the  `VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT` set and it does an implicit CB reset a number of times because BeginCommandBuffer is called in a loop for each subtest.  The GPU tracking state remains intact through these resets.

ResetCommandPool tells the implementation to recycle command buffer resources.  This doesn't mean that the layers have to recycle tracking info.  The command buffers are put back into the initial state; they are not destroyed or recreated.  The GPU validation resources attached to a CB have the same life cycle as the CB as bracketed by Allocate and Free CB.

karl-lunarg commented:
Rebased and pushed:

- 3rd item from the above list
- picked up updated known-good from master, so "known-good" commits no longer needed

I'm leaving the "Track ShaderModule" commit un-squashed because it fixes a more general state tracking problem that was exposed by gpu-validation.

karl-lunarg commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,670 @@
+
+Here are the options related to activating GPU-Assisted Validation:
+
+1. enable - Enables GPU-Assisted Validation
```
This section was intended to be a description of the options that is independent of how they are activated.  The activation process is different between the layer settings file and the programmatic interface, with slightly different terms and names.  These details are in the following sections.  I've reworded this "generic" description so that it is not easily confused with the specifics described below.

karl-lunarg commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,670 @@
+    Use the instrumented bytecode to create the ShaderModule.
+* After calling QueueSubmit, perform a wait on the queue to allow the queue to finish executing.
+    Then map and examine the device memory block for each command buffer that was submitted.
+    If any debug record is found, generate a validation error message.
```
fixed

karl-lunarg commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,670 @@
+Some platforms do not support configuration of the validation layers with this configuration file.
+Programs running on these platforms must then use the programmatic interface.
+
+### Enabling With the Programmatic Interface
```
It appears in the enum type in the header (or will soon) which implies that it can be used.  But I've added a note that it can be used here.

As an aside, this isn't intended to be end-user/SDK-user documentation.  It is more of an internal "how this works" sort of thing.  There are other user-facing docs that will be enhanced with the appropriate treatment of part of this material.

Tony-LunarG commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    if (new_create_info.setLayoutCount >= gpu_state->adjusted_max_desc_sets) {
+        std::ostringstream strm;
+        strm << "Pipeline Layout conflict with validation's descriptor set at slot " << gpu_state->desc_set_bind_index << ". "
+             << "Application has too many descriptor sets in the pipeline layout. "
```
too many descriptor sets in the pipeline layout  to continue with gpu validation

Tony-LunarG commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,669 @@
+
+#### GpuPostCallRecordAllocateCommandBuffers
+
+* Get a descriptor set for each command buffer from the descriptor set manager
```
Could this go in the "For each command buffer" list below?

karl-lunarg replied:
Oh, well, the descriptor set manager allows the caller to get a bunch of descriptor sets in one call.  So if the application called AllocateCommandBuffers for more than one command buffer, this function would make only one call to get the descriptors for them all.  So this is a case of this optimization creeping into the design, so I'll reword it as suggested.

Tony-LunarG commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,669 @@
+
+  This should only need to be done after making any major changes to the implementation.
+
+* The additional API calls are not state-tracked
```
But wouldn't they be state tracked using the double core_validation method above?  Or is this bullet independent of the above discussion.

karl-lunarg replied:
They are separate discussions.  You wouldn't normally run in the double core method and so the additional API calls are not normally tracked.  The double core method is just a suggested technique for testing the GPU validation code on an occasional basis or when suspecting that there is something wrong with it.

karl-lunarg replied:
I'll go ahead and resolve this with no changes.  Let me know if you have questions or still would like to see something different here.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -5064,6 +5143,9 @@ static void PreCallRecordDestroyPipeline(layer_data *dev_data, VkPipeline pipeli
     InvalidateCommandBuffers(dev_data, pipeline_state->cb_bindings, obj_struct);
     dev_data->pipelineMap.erase(pipeline);
+    if (GetEnables(dev_data)->gpu_validation) {
+        GpuPreCallRecordDestroyPipeline(dev_data, pipeline_state);
```
pipeline_state is stale here as the erase has deleted it (it's a map of unique pointers).  Move above erase or copy out (before erase) the pipeline_state->pipeline field and pass as the argument (the only one `GpuPreCallRecordDestroyPipeline` uses).

karl-lunarg replied:
Fixed.  Thank you.  I'm now passing only the handle to `GpuPreCallRecordDestroyPipeline`.

I also moved the erase to after the call to the Gpu function because it makes sense to do the erase last, in case anything is depending on it before it is destroyed.  This also makes it a little less likely to make the same mistake again.

jzulauf-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            block.buffer = cb_node->gpu_output_buffer;
+            block.memory = cb_node->gpu_output_memory;
+            block.offset = cb_node->gpu_output_memory_offset;
+            gpu_state->memory_manager->PutBackBlock(block);
```
code would be more readable is PutBackBlock had an overload that took buffer, memory, offset as three args -- or if cb_node stored the gpu_output_buffer et. seq. as a  `GpuDeviceMemoryManager::GpuDeviceMemoryBlock`

karl-lunarg replied:
Agreed.  I added the overload for PutBackBlock.

Fixed.

jzulauf-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            block.memory = cb_node->gpu_output_memory;
+            block.offset = cb_node->gpu_output_memory_offset;
+            gpu_state->memory_manager->PutBackBlock(block);
+            cb_node->gpu_output_buffer = VK_NULL_HANDLE;
```
if this was a `GpuDeviceMemoryManager::GpuDeviceMemoryBlock` a reset method (or assignment to default value) would be tidy.

karl-lunarg replied:
Agreed as well.   But then I'd have to expose the `GpuDeviceMemoryManager::GpuDeviceMemoryBlock` definition outside of the file containing most of the gpu validation code.  And I'm sort of leaning towards keeping that contained and also not sticking foreign objects in cb_node, which might hurt a little.

Part of all this stems from the current discomfort of a lot of definitions being in core_validation.cpp and code that is part of core validation but happens to be in another file (shader_validation.cpp, gpu_validation.cpp, etc) has sort of a second-class degree of visibility to these things as compared to code in core_validation.cpp.  I'm thinking this will get cleaned up someday, and I don't want to add to the mess in return for a small readability gain.

jzulauf-lunarg replied:
It's a useful enough struct you could move the definition up a level (to core_validation_types)

karl-lunarg replied:
OK, done

tobine commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+
+    // The descriptor indexing checks require only the first "output" descriptor set.
+    // The descriptor initialization checks require inputs for each stage.
+    const VkDescriptorSetLayoutBinding debug_desc_layout_bindings[6] = {
```
Rather than hard-coded "6" would prefer const variable kNumShaderStages or something like that.

karl-lunarg replied:
This turned out to be better advice than it first appeared because I had "6" hard-coded in a few places that would have been easy to forget about.  Thanks!

Fixed.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2300,6 +2303,25 @@ bool OutsideRenderPass(const layer_data *dev_data, GLOBAL_CB_NODE *pCB, const ch
+        CORE_VALIDATION_GPU_VALIDATION_RESERVE_BINDING_SLOT_BIT = 0x00000002,
+    };
+    typedef VkFlags CoreGPUFlags;
+    const std::unordered_map<std::string, VkFlags> gpu_flags_option_definitions = {
```
fixed

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2313,11 +2335,44 @@ void SetDisabledFlags(instance_layer_data *instance_data, const VkValidationFlag
                 // Set all disabled flags to true
                 instance_data->disabled.SetAll(true);
                 break;
+            case 2:  // TODO awaiting spec update - temp backdoor
```
> Prefer a symbolic name...

This was just a hack that I think only I used from a VLT to confirm that I could turn on gpu validation from a VLT.  Once the spec work for `VK_EXT_validation_features` got far enough along so I could paste the pending header file definitions in, the need for this backdoor vanished.

I'll simply remove it now.

> deprecation warning

The spec text for `VK_EXT_validation_features` uses the word "subsume" to describe the relationship to `VK_EXT_validation_flags`.  I'd prefer to not add a message like this until we start seeing the word "deprecated" associated with `VK_EXT_validation_flags` in the spec.

If you are talking about a message for the "2" backdoor, no, no one else should be using it.

mark-lunarg commented on tests/vkrenderframework.cpp:
```diff
@@ -298,6 +298,14 @@ void VkRenderFramework::GetPhysicalDeviceProperties(VkPhysicalDeviceProperties *
     *props = vk_testing::PhysicalDevice(gpu()).properties();
 }
 
+bool VkRenderFramework::DeviceIsMockICD() {
```
I need this awesome function to push the LX817 fix.  Hope it goes in soon!

karl-lunarg replied:
I'm tryin'

(We do have the option of cherry-picking the Framework commit into master if you need it sooner.  But I suspect this will merge to master soon after the next Vulkan header gets integrated.)

One thing to be aware of is that if the devsim layer is around, it can potentially change the `VkPhysicalDeviceProperties'` `vendorID`, `deviceID`, and `deviceName` members.
The device profiles that we run in the VLT on Travis-CI do this.  This causes this function to think that it is not the MockICD, even though it is.  And I found that out the hard way.

Look at the VLT test in this PR to see that I have to combine the check for the MockICD with a check for the devsim layer.

This is a little unfortunate, but I understand why people would want to override these fields with devsim.  However, this doesn't really give us a good way to detect the MockICD.  It would nice if we could come up with a better way.

karl-lunarg replied:
I suppose we could register a dummy extension with Khronos, saying that it should be implemented only by the mock ICD.  Then clients could look for that extension to detect the mock ICD.  But I'm not sure that devsim is or ever will be capable of changing a device's extension list.  If that happens, this could be less reliable as well.

@tobine, any thoughts?

chrisforbes replied:
I'm not sure about the paranoia about the devsim layer? It's only ever been safe to lower limits / disable things; how is this different?

karl-lunarg replied:
@chrisforbes devsim can be used to change the "name" of the device via the `vendorID`, `deviceID`, and `deviceName` properties.  This makes it harder to tell if the device is the Mock ICD or not.

The VLT for gpu validation actually needs to run a shader in order to verify operation.  The Mock ICD can't do that, among many other things.  So I need to not run the VLT for gpu validation when the device isn't "real", meaning the device is the Mock ICD.  If devsim is around, then I can't be sure it is the Mock ICD or not.

One can entertain the possibility that devsim should not be allowed to change the identification of the device (these three properties) and just be used to lower limits and disable things.   But in the context of the VLT testing on Travis-CI, devsim is used to "create" a number of devices, using the MockICD as the "common base".  If all the devsim device definitions used in these VLT Travis-CI tests did NOT alter the device naming properties, leaving the device to identify itself as the MockICD, then we wouldn't have this problem.

After thinking about this some more, I sort of like the idea of devsim not altering the device identification fields.  We can change the device devsim definitions to not do this, which solves the immediate problem for VLT. Or we can change devsim to not allow changing these fields.  The current usage is outright lying about the device and that's why we're talking about this.

Another way to handle this is to test for a feature (like "can run shader program") that is on real devices and not on fake ones.  I haven't been able to find such a feature and that's why I suggested the dummy MockICD-only extension as a means of reliably detecting the MockICD.

Bottom line, I think that the correct approach is to NOT change devsim to prohibit alter device identification fields.  There are some valid use cases.  For example, you might be running on a real device, but you have some behavior that keys off the `deviceID` and it may be valid to tweak that `deviceID` to another "close" device that is still OK on the real device.

But what we should do is remove the device identification overrides in the devsim profiles we're running on Travis-CI.  I don't think that removing these is going to hurt anything.  The Travis-CI logs identify the device that is being used by the filename of the devsim device definition:

    Testing with profile tests/device_profiles/geforce_1080_ti.json

So the test case identification is via this filename and not via the device identification fields in the profile anyway.  Since the device profiles seem to have been pulled from gpuinfo.org, I'm guessing that the device identification fields were not intentionally added anyway.

Still would like to see @tobine's opinion.

tobine replied:
For VLT, certainly, there is no need to be changing the vendorID, deviceID, or deviceName. That's just a side effect of the fact that the profiles were downloaded gpuinfo.org as-is an not modified. So certainly a simple/straightforward solution for VLT is to just update the device profile files we use to remove those fields.
A slightly heavier weight solution would be making it impossible for devsim to modify those IDs. That seems potentially against the spirit of devsim.
Adding a mock-ICD-only extension is not a bad idea, but is probably more than we need as I suspect that just updating the device profiles used for testing is sufficient.

chrisforbes replied:
Another approach here would be to submit some trivial (correct) gpu work, see if the driver actually does it.

karl-lunarg replied:
I thought that this would be too much to do now, but I went ahead and implemented a framework function to submit gpu work to see if it draws.  The result was pretty encouraging.  And I think that   its new supporting functions can be used to read memory after a draw to inspect the results for other testing purposes.

I'll pull it together in a new PR and try to put it up later this morning.  I'll update the VLT to use it once it hits master and I rebase this branch.

In short, there will be a new framework function `DeviceCanDraw()`.  It calls another new framework function that copies the render target's image to another "staging" image with the appropriate layout.  Finally, it calls another new function that copies the pixels out of the staging image into host memory.  It then checks to see if the expected contents are there.  `DeviceCanDraw()` needs to be called before `Init()`, but that's fine and even preferable for this use case.

I'll probably retain the simple `DeviceIsMockICD` function as well, but I'll document that devsim can confuse it.

karl-lunarg replied:
This is all sorted now, as described above.  The new framework support has been merged to master and I've rebased and updated the VLT to use it.

Other VLT tests are already using the new support to avoid running on fake devices.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -13846,6 +13990,21 @@ VKAPI_ATTR VkResult VKAPI_CALL EnumeratePhysicalDevices(VkInstance instance, uin
+    instance_layer_data *instance_data = GetLayerDataPtr(get_dispatch_key(physicalDevice), instance_layer_data_map);
+    instance_data->dispatch_table.GetPhysicalDeviceProperties(physicalDevice, pPhysicalDeviceProperties);
+    if (instance_data->enabled.gpu_validation && instance_data->enabled.gpu_validation_reserve_binding_slot) {
+        if (pPhysicalDeviceProperties->limits.maxBoundDescriptorSets > 1) {
```
It is a bit of defensive coding because returning a zero here would be bad.  I did look at gpuinfo.org and saw some devices there that had only 1.  I realize that these would be non-conforming drivers, but that doesn't mean that they won't be used.

chrisforbes replied:
It would be nice to make this choice in one place rather than two, though?

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+        block_size_ = ((record_size_ + align - 1) / align) * align;
+        blocks_per_chunk_ = kItemsPerChunk;
+        chunk_size_ = blocks_per_chunk_ * block_size_;
+        this->dev_data_ = dev_data;
```
I don't know.  It is possible that at one point I didn't name the class members with a trailing '_' and so I needed to do that for this line.

Fixed.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    const VkDescriptorSetLayoutCreateInfo dummy_desc_layout_info = {VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO, NULL, 0, 0,
+                                                                    NULL};
+
+    VkResult result = GetDispatchTable(dev_data)->CreateDescriptorSetLayout(GetDevice(dev_data), &debug_desc_layout_info, NULL,
```
I did this here and in a couple of other functions where it made sense.

Fixed.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    void PutBackBlock(GpuDeviceMemoryBlock &block) {
+        // Find the chunk belonging to the allocated offset and make the offset available again
+        auto chunk = std::find_if(std::begin(chunk_list_), std::end(chunk_list_), [&block](const MemoryChunk &c) {
+            return (c.buffer == block.buffer) && (c.memory == block.memory);
```
There's really no good reason.  I'll take out the second check.

Fixed.

chrisforbes commented on layers/core_validation.cpp:
```diff
@@ -10955,11 +10955,10 @@ static bool CreatePassDAG(const layer_data *dev_data, RenderPassCreateVersion rp
                                              VkShaderModule *pShaderModule, uint32_t unique_shader_id) {
     spv_target_env spirv_environment = ((GetApiVersion(dev_data) >= VK_API_VERSION_1_1) ? SPV_ENV_VULKAN_1_1 : SPV_ENV_VULKAN_1_0);
-    unique_ptr<shader_module> new_shader_module(
-        spirv_valid ? new shader_module(pCreateInfo, *pShaderModule, spirv_environment, unique_shader_id) : new shader_module());
```
How does this interact with NV_glsl_shader? "shaders that aren't spirv at all" was the original motivation for this.

karl-lunarg replied:
Ugh, then I'll have to think of something more sophisticated to deal with this case.  And also handling non-SPIRV shaders in other parts of all this.

karl-lunarg replied:
I widened the interface to `PreCallValidateCreateShaderModule` to add an `is_spirv` boolean which just returns true if the magic number is present.

I then put back `PostCallRecordCreateShaderModule` pretty much the way it was, except that I create a non-empty shader_module if `is_spirv` is true.  This changes the criteria for creating a non-empty shader_module to if it is SPIR-V, valid or not.

Some applications want to create the shader anyway, even if the SPIR-V is invalid.  They do this by returning skip=false in the debug callback if shader validation fails.

I'll have to add a couple of more checks to avoid trying to instrument shaders that are not SPIR-V, etc.

jzulauf-lunarg commented:
@karl-lunarg -- the "dispatch while locked" comments don't need to be fixed for this initial commit as long as they are limited to the GPU validation enabled path.  This issue should be addressed in a subsequent PR, but it shouldn't gate this one.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -5632,6 +5727,12 @@ const DeviceFeatures *GetEnabledFeatures(const layer_data *device_data) { return
+GpuValidationState *GetGpuValidationState(layer_data *device_data) { return &device_data->gpu_validation_state; }
+const GpuValidationState *GetGpuValidationState(const layer_data *device_data) { return &device_data->gpu_validation_state; }
+
+VkDevice GetDevice(layer_data *device_data) { return device_data->device; }
```
I'll address the "non-const argument" part of this comment.  I don't understand the "by value" or  the "any point to a const returning one" parts of the comment.

Most of the Gpu* functions in gpu_validation.cpp take the `layer_data *dev_data` argument as non-const because they need to modify GPU Validation state, which lives in the layer_data.

Note that most of the functions in a file like buffer_validation.cpp do the same thing because they modify other state in layer_data.

What makes gpu_validation.cpp a bit different is that it needs the VkDevice handle because it makes additional "API" calls via the device dispatch table and needs to pass a VkDevice as (often) the first argument.  We can't just use `dev_data->device` in gpu_validation.cpp because the layer_data struct lives for some odd reason in core_validation.cpp.  It is an incomplete type from the point of view of gpu_validation.cpp.  This basically means that functions outside of core_validation.cpp need to get at layer_data elements via accessor functions like GetDevice.  I have something similar for solving the same problem for getting the GPU Validation state (GetGpuValidationState).

Accessor functions are great, especially when you want to express constness and start thinking about distinguishing between read and write locking.  But here, I need both GetDevice functions because I get the layer_data as const and non-const.

If we're that allergic to the GetDevice with non-const arg, I could put a copy of the VkDevice handle in my GpuValidationState and access it from there.  But then, we'd have two copies of the handle in layer_data which is also distasteful.

No plan to change anything here.

karl-lunarg replied:
ok, fixed.

jzulauf-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    // descriptor set index.
+    std::vector<unsigned int> pipeline_uses_debug_index(count);
+    for (uint32_t pipeline = 0; pipeline < count; ++pipeline) {
+        for (uint32_t stage = 0; stage < pCreateInfos[pipeline].stageCount; ++stage) {
```
Reconstructing this information seems redundant.

This information is in the recorded PIPELINE_STATE::active_slots as the key, populated in PreCallValidateCreateGraphicsPipelines among other places.

If that isn't useful the code to create that information is in CollectInterfaceByDescriptorSlot

chrisforbes commented on layers/core_validation_types.h:
```diff
@@ -1134,6 +1152,20 @@ struct DeviceFeatures {
+    uint32_t adjusted_max_desc_sets;
+    uint32_t desc_set_bind_index;
+    uint32_t unique_shader_module_id;
+    std::vector<std::tuple<VkShaderModule, uint32_t, VkPipeline, std::vector<unsigned int>>> shader_map;
```
This is amazingly opaque; I have to read the rest of your change to have any idea what this does. Describe here?

karl-lunarg replied:
Changed to use an unordered_map.
Fixed.

chrisforbes commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,669 @@
+### Descriptor Set Binding Limit
+
+This is probably the most important limitation and is related to the
+`VkPhysicalDeviceLimits::maxBoundDescriptorSets` device limit.
```
This approach is only guaranteed to work for compute shaders.

For graphics shaders, you need to check for (and enable if the app doesn't) the VkPhysicalDeviceFeatures::fragmentStoresAndAtomics and VkPhysicalDeviceFeatures::vertexPipelineStoresAndAtomics feature bits.

And this is likely where the disappointment starts -- most mobile devices that support Vulkan do not expose these features.

karl-lunarg replied:
I've added code to enable these features if supported.  I'll update docs to reflect this as well.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            return (c.buffer == block.buffer) && (c.memory == block.memory);
+        });
+        if (chunk_list_.end() == chunk) {
+            assert(true);
```
false, presumably.

karl-lunarg replied:
yes, fixed.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            chunk->available_offsets.push_back(block.offset);
+            if (chunk->available_offsets.size() == blocks_per_chunk_) {
+                // All offsets have been returned
+                FreeMemoryChunk(*chunk);
```
Considered some hysteresis?

karl-lunarg replied:
Yes, I understand the need and did consider it.  But I needed to get on with things...

What is a good way to do this?   Put the chunk on a deferred free list and if the chunk is still unused after some number of GetBlock requests, then free it?

karl-lunarg replied:
Oh, perhaps figure out how many free blocks there are across the other chunks and free this chunk only if there is a certain amount?  Something like that should work.

karl-lunarg replied:
I need to defer this.  No change planned.  At worst, this would be a rare perf problem.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            }
+            const VkDescriptorPoolSize size_counts = {
+                VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
+                pool_count * 6,
```
Why 6?

karl-lunarg replied:
I replaced this 6 (and some others) with a constant, in response to a comment @tobine made.
Already fixed (put not pushed yet)

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+
+    VkResult GetDescriptorSets(uint32_t count, VkDescriptorPool *pool, std::vector<VkDescriptorSet> &desc_sets) {
+        auto gpu_state = GetGpuValidationState(dev_data_);
+        const uint32_t default_pool_size = kItemsPerChunk;
```
I suspect you're going to want a separate tunable value for this...

karl-lunarg replied:
That was part of the thinking.  There is some difficulty in exposing a tunable value via the API because of the need to add it somehow to an extension.  It is pretty easy to define an extension like `VK_EXT_validation_features` because you can come along later and add additional enable/disable enums to it without breaking anything.  It is much harder to define such an extension that has a struct of values for example, because you can't change that struct, or even add on to it.

In any case, this implementation of a descriptor set manager uses multiple pools, so it will allocate/free pools as needed.  So, at worst, not exposing this as tunable would be a perf issue and not a hard ceiling on the pool size.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    };
+
+    layer_data *dev_data_;
+    std::list<struct PoolTracker> desc_pool_list_;
```
unordered, and partitioned by emptyness, perhaps?

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+
+    // Some devices have extremely high limits here, so set a reasonable max because we have to pad
+    // the pipeline layout with dummy descriptor set layouts.
+    gpu_state->adjusted_max_desc_sets = GetPhysicalDeviceProperties(dev_data)->limits.maxBoundDescriptorSets;
```
As in docs comment -- you need to ensure you can actually write to memory from the shader stages you want.

chrisforbes replied:
On capable devices, but where the app doesn't enable features it doesn't use, this will also involve modifying the features requested, before GpuPostCallRecord...

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            2,
+            VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
+            1,
+            VK_SHADER_STAGE_TESSELLATION_CONTROL_BIT,
```
Is it OK to mention the bits for these shader stages on implementations where they don't exist? In most cases this is a VU problem.

karl-lunarg replied:
I'm going to take out bindings 1-5, leaving binding 0.

Bindings 1-5 are for the second phase of this effort and I put them in here as a placeholder for that phase.  The second phase uses these bindings to pass information TO the instrumented shader code to communicate which descriptors the application has updated.  This is to support the "partial bind" and the "variable descriptor count" features of the `VK_EXT_descriptor_indexing` extension.

These bindings serve no purpose for this first phase.  Taking them out will reduce some confusion and reduce pressure on the descriptor pools.

chrisforbes commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,669 @@
+This is probably the most important limitation and is related to the
+`VkPhysicalDeviceLimits::maxBoundDescriptorSets` device limit.
+
+Basically, when applications use all the available descriptor set binding slots,
```
What do you want to do about other limits -- maxFragmentCombinedOutputResources, etc. Your debug SSBO in that stage counts against that limit, and may conflict with MRT setups.

chrisforbes commented on layers/core_validation.cpp:
```diff
@@ -10828,26 +10955,43 @@ static bool CreatePassDAG(const layer_data *dev_data, RenderPassCreateVersion rp
+    if (GetEnables(dev_data)->gpu_validation) {
+        VkShaderModuleCreateInfo instrumented_create_info = *pCreateInfo;
+        std::vector<unsigned int> instrumented_pgm;
+        bool pass = GpuInstrumentShader(dev_data, pCreateInfo, instrumented_pgm, &unique_shader_id);
```
How does this work when a shader module contains multiple entrypoints?

(both the basic case, where you consume one of them; and the complex case where multiple stages in the pipeline use the same shader module, but different entrypoints?)

karl-lunarg replied:
@greg-lunarg indicates that this won't work for shaders with multiple entry points, partly because the instrumentation code needs to know what stage is running it. I believe that the instrumentation is just skipped for such shaders which means that they won't report any errors.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    if ((result != VK_SUCCESS) || (result2 != VK_SUCCESS)) {
+        ReportSetupProblem(dev_data, VK_DEBUG_REPORT_OBJECT_TYPE_DEVICE_EXT, HandleToUint64(GetDevice(dev_data)),
+                           "Unable to create descriptor set layout.  GPU-Assisted Validation disabled.");
+        delete gpu_state->desc_set_manager;
```
Hold these by unique_ptr ?

karl-lunarg replied:
I took an honest shot at this and got most of the way there.  The problem is that I have to store the unique_ptr in the device state and the layer_data destructor doesn't have access to a destructor for this class.  I'd have to expose the entire class in core_validation_types, which is more than I want to do right now.  I know the current implementation is a bit C-ish.

karl-lunarg replied:
I changed my mind on this and went ahead and exposed the class definitions for both manager classes via the gpu_validation.h file.  That let me track the classes with unique_ptrs in the GPU validation state.

This will also let me take care of some of the other comments from @jzulauf-lunarg more easily.

Things are a bit cleaner now.  And the manager destructors are automatically getting deleted when the device data goes away at destroy device time.  So that's cool.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+                create_info.pCode = shader->words.data();
+                create_info.codeSize = shader->words.size() * sizeof(uint32_t);
+                VkResult result =
+                    GetDispatchTable(dev_data)->CreateShaderModule(GetDevice(dev_data), &create_info, NULL, &shader_module);
```
Consider using the allocator passed in?

karl-lunarg replied:
Yes.
Fixed.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    // Call the optimizer to instrument the shader.
+    // Use the unique_shader_module_id as a shader ID so we can look up its handle later in the shader_map.
+    using namespace spvtools;
+    spv_target_env target_env = SPV_ENV_UNIVERSAL_1_2;
```
This is probably not the right environment?

karl-lunarg replied:
After consulting with @greg-lunarg:

* Yes, we should use `SPV_ENV_VULKAN_1_1` here.
* The gpu instrumentation code uses SPIR-V 1.3 features, which requires a Vulkan 1.1 implementation.  So I'm adding a check in PostCreateDevice that turns off gpu validation if the device is < 1.1.

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+                            VkFence fence) {
+    auto gpu_state = GetGpuValidationState(dev_data);
+    if (gpu_state->aborted) return;
+    for (uint32_t submit_idx = 0; submit_idx < submitCount; submit_idx++) {
```
Sync behavior would be clearer if you put the QWI in this function?

chrisforbes replied:
Don't you also want a barrier that makes the debug data visible to the host? The QWI by itself doesn't seem adequate.

jzulauf-lunarg replied:
From the spec:

> vkQueueWaitIdle is equivalent to submitting a fence to a queue and waiting with an infinite timeout
for that fence to signal

From vkWaitForFences

> Signaling a fence and waiting on the host does not guarantee that the results of
memory accesses will be visible to the host, as the access scope of a memory
dependency defined by a fence only includes device access. A memory barrier or
other memory dependency must be used to guarantee this. See the description of
host access types for more information.

karl-lunarg replied:
Right now, the buffer memory is allocated as host coherent.  The spec says:

> VK_MEMORY_PROPERTY_HOST_COHERENT_BIT bit specifies that the host cache management commands vkFlushMappedMemoryRanges and vkInvalidateMappedMemoryRanges are not needed to flush host writes to the device or make device writes visible to the host, respectively.

So we are guaranteeing _availability_ by the QWI, meaning that the submission has been completed and the result is "ready", but perhaps not visible through the caches, etc.

However, _visibility_ is guaranteed by the coherency status.  In particular, your phrase from WaitForFences:

> A memory barrier or other _memory dependency_ must be used to guarantee this.

is satisfied by the fact that the memory is coherent.  In particular:

>not needed to ... _make device writes visible to the host_

from the coherency description satisfies the dependency.

We recognize that HOST_COHERENT is not available on all platforms.  I plan to add code now to call `vkInvalidateMappedMemoryRanges` if we can't get HOST_COHERENT.

But I don't think we need a barrier.

Our tests are passing on NVIDIA and Intel graphics.  However, for AMD, we're seeing only the first word written back by the GPU, which is under (mostly) separate investigation.  While it seems odd that only the first word is visible, the first word in the debug buffer is used for a very different purpose than the other words comprising the error records.  See the design doc for more info.

Anyway, we're continuing to dig into this and are not completely ruling out using barriers.  But everything we've tried to date, including using very aggressive barriers has had no effect.

karl-lunarg replied:
oops, HOST_COHERENT is required, so handling the case without it is not needed.

chrisforbes replied:
1/ Coherent memory isn't adequate. You still need a domain operation to make these device writes available to the host domain.

Spec (note that this is host->host domain, host domain->host only): 

> If the memory object does have the VK_MEMORY_PROPERTY_HOST_COHERENT_BIT property flag, writes to the memory object from the host are automatically made available to the host domain. Similarly, writes made available to the host domain are automatically made visible to the host.

2/ It's true that there must be a heap with HOST_COHERENT. It's _not_ true that your storage buffers must be able to use it.

karl-lunarg replied:
I wasn't _completely_ convinced that the barrier (domain operation) wasn't necessary.  Looking at diagrams, I thought it strange how a host bus transaction would provoke any unavailable device writes to become available.

It may be the case that all the anecdotal "evidence" and interpretations around don't take into account that some apps might be issuing barriers or be issuing other operations that would make this device data available.

In any case, this PR might get merged to master anyway shortly to start the SDK process.  We'll do our best to get any fixes for these issues in the SDK.

chrisforbes replied:
If you want to be done with it quickly, throw in a full barrier (ALL_COMMANDS_BIT, MEMORY_WRITE BIT) -> (HOST_BIT, HOST_READ_BIT) here. You're already radically oversynchronizing by forcing the queue idle after every submission; a little more to get the memory dependencies right isn't going to hurt.

karl-lunarg replied:
Yes, I already have some code set off in another branch that does exactly this, which we used to initially address the problem.  I should be able to incorporate it back in pretty easily.

karl-lunarg replied:
> 2/ It's true that there must be a heap with HOST_COHERENT. It's not true that your storage buffers must be able to use it.

How is this determined?

chrisforbes commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+void GpuPreCallRecordDestroyPipeline(layer_data *dev_data, const PIPELINE_STATE *pipeline_state) {
+    auto gpu_state = GetGpuValidationState(dev_data);
+    for (auto it = gpu_state->shader_map.begin(); it != gpu_state->shader_map.end();) {
+        if (std::get<2>(*it) == pipeline_state->pipeline) {
```
It seems this wants to be something more like an unordered_map keyed on the VkPipeline?

karl-lunarg replied:
Yes, it is on my list to address this.

karl-lunarg replied:
Actually, no.  We can't use VkPipeline as a key, because there can be more than one shader module per VkPipeline.  I can use the VkShaderModule or the unique shader ID as a key.  The unique shader ID is probably better because I never use the VkShaderModule to look up anything.

karl-lunarg replied:
OK, changed to use unordered_map keyed by the unique ID, which works well.  Docs updated too.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2495,7 +2556,9 @@ static void PreCallRecordCreateDevice(VkLayerDeviceCreateInfo *chain_info) {
 }
 
-static void PostCallValidateCreateDevice(const VkDeviceCreateInfo *pCreateInfo) { ValidateLayerOrdering(*pCreateInfo); }
+static void PostCallValidateCreateDevice(const VkDeviceCreateInfo *pCreateInfo, VkDevice *pDevice) {
```
I thought I typed in a response when I marked this resolved.  This is fixed now.  I think I verified that I had an earlier version that used this parameter and forgot to remove it when I removed the need for it.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+            desc_pool_list_.push_front(tracker);
+            pool_to_use = &desc_pool_list_.front();
+        }
+        std::vector<VkDescriptorSetLayout> desc_layouts;
```
yes, fixed.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+        uint32_t offset;
+    };
+
+    VkResult GetBlock(GpuDeviceMemoryBlock *block) {
```
I wanted the VkResult to be the return value, even though I'm not doing that much with it now.

I added the suggested assert and cleared the block before the call in the one place this is used.

Feel free to spif this up more later if you wish.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    }
+
+    void PutBackDescriptorSet(VkDescriptorPool desc_pool, VkDescriptorSet desc_set) {
+        for (std::list<PoolTracker>::iterator pool = desc_pool_list_.begin(); pool != desc_pool_list_.end(); ++pool) {
```
This has been changed to use a map.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+        desc_sets.clear();
+        desc_sets.resize(count);
+
+        for (auto &pool : desc_pool_list_) {
```
I don't think it is worth it.
No change.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -7270,6 +7389,11 @@ VKAPI_ATTR void VKAPI_CALL CmdBindPipeline(VkCommandBuffer commandBuffer, VkPipe
     if (!skip) dev_data->dispatch_table.CmdBindPipeline(commandBuffer, pipelineBindPoint, pipeline);
+    if (GetEnables(dev_data)->gpu_validation) {
+        lock.lock();
+        GpuPostCallRecordCmdBindPipeline(dev_data, commandBuffer, pipelineBindPoint, pipeline);
```
Perhaps I tried too hard to stick with the established conventions.  But this is a different animal.  I've done some renaming of the hooked functions that deviate from the convention along the lines you suggest.  I also added the suggested comment.
Fixed.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -9823,6 +9947,9 @@ static void PreCallRecordCmdWaitEvents(layer_data *dev_data, GLOBAL_CB_NODE *cb_
         [=](VkQueue q) { return ValidateEventStageMask(q, cb_state, eventCount, first_event_index, sourceStageMask); });
     TransitionImageLayouts(dev_data, cb_state, imageMemoryBarrierCount, pImageMemoryBarriers);
+    if (GetEnables(dev_data)->gpu_validation) {
+        GpuPreCallRecordCmdWaitEvents(dev_data, sourceStageMask);
```
I renamed to validate which is slightly better.
Fixed.

karl-lunarg commented on layers/core_validation_types.h:
```diff
@@ -1020,6 +1023,12 @@ struct GLOBAL_CB_NODE : public BASE_NODE {
+    // GPU Validation data
+    VkBuffer gpu_output_buffer;
+    VkDeviceMemory gpu_output_memory;
+    uint32_t gpu_output_memory_offset;
```
Fixed as indicated in other comments.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+}
+
+// Just gives a warning about a possible deadlock.
+void GpuPreCallRecordCmdWaitEvents(layer_data *dev_data, VkPipelineStageFlags sourceStageMask) {
```
Yes, I renamed it to validate which is only slightly better.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -10828,26 +10955,43 @@ static bool CreatePassDAG(const layer_data *dev_data, RenderPassCreateVersion rp
     if (PreCallValidateCreateShaderModule(dev_data, pCreateInfo, &spirv_valid)) return VK_ERROR_VALIDATION_FAILED_EXT;
 
-    VkResult res = dev_data->dispatch_table.CreateShaderModule(device, pCreateInfo, pAllocator, pShaderModule);
+    if (GetEnables(dev_data)->gpu_validation) {
```
ok, done

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+        desc_pool_list_.clear();
+    }
+
+    VkResult GetDescriptorSets(uint32_t count, VkDescriptorPool *pool, std::vector<VkDescriptorSet> &desc_sets) {
```
OK, fixed.

mark-lunarg commented on layers/core_validation.cpp:
```diff
@@ -7296,6 +7402,12 @@ VKAPI_ATTR void VKAPI_CALL CmdBindPipeline(VkCommandBuffer commandBuffer, VkPipe
     }
     lock.unlock();
     if (!skip) dev_data->dispatch_table.CmdBindPipeline(commandBuffer, pipelineBindPoint, pipeline);
+    if (GetEnables(dev_data)->gpu_validation) {
```
PostCallRecord function?

karl-lunarg replied:
There was another (closed) conversation discussing the names of these functions.  The gist of it was that "PostCallRecord" is misleading here because it doesn't record anything and in fact dispatches another command entirely.

I did try it "your way" by using the "closest" existing naming convention, but this gpu validation thing is a different animal and forcing things into this convention doesn't work and is confusing, at least to other reviewers.  I ended up agreeing.
No change.

mark-lunarg replied:
It is a bit confusing (maybe PostCallRecord is to restrictive?) , but it ultimately will happen in some fashion and doesn't violate the convention.

mark-lunarg commented on layers/core_validation.cpp:
```diff
@@ -15712,6 +15848,7 @@ static const std::unordered_map<std::string, void *> name_to_funcptr_map = {
     {"vkCreateInstance", (void *)CreateInstance},
     {"vkCreateDevice", (void *)CreateDevice},
     {"vkEnumeratePhysicalDevices", (void *)EnumeratePhysicalDevices},
+    {"vkGetPhysicalDeviceProperties", (void *)GetPhysicalDevicePropertiesIntercept},
```
Why *Intercept?

karl-lunarg replied:
Are you suggesting just removing the "Intercept"?

mark-lunarg replied:
That's what I'd suggest.

karl-lunarg replied:
`GetPhysicalDeviceProperties()` is an existing "helper" function that returns cached physical device properties that were saved in the device as part of `PostCallRecordCreateDevice()`  This helper function is used by buffer_validation.cpp. 

The layer never intercepted `vkGetPhysicalDeviceProperties()`.  I have to add the intercept in order to adjust one of the limits for gpu validation.  The name "GetPhysicalDeviceProperties" was already taken.  So I needed something different.  Let me know if you have a better name.

No change planned.

mark-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2300,6 +2303,25 @@ bool OutsideRenderPass(const layer_data *dev_data, GLOBAL_CB_NODE *pCB, const ch
 static void InitCoreValidation(instance_layer_data *instance_data, const VkAllocationCallbacks *pAllocator) {
```
This function is going away Very Soon™ and should go in the CreateInstance pre- or post-call record function, but we can leave it until conflict resolution time if it comes to that.

karl-lunarg replied:
I then trust that whoever does the conflict resolution does it correctly.

Tony-LunarG commented on layers/gpu_validation.h:
```diff
@@ -0,0 +1,135 @@
+ * limitations under the License.
+ *
+ */
+
```
I seem to need an "include <algorithm>" here to build on Windows to avoid a "error C2039: 'max': is not a member of 'std'"

Tony-LunarG replied:
That should be \<algorithm\>

karl-lunarg replied:
There's also the NOMINMAX thing on Windows too.  I've fixed this, but by removing the one trivial usage of max in the header file.  This problem resulted from moving the class defs into the header.  Since the NOMINMAX thing has order dependencies with Windows.h, I didn't want to introduce an order dependency in other files that include this header.

I don't push every commit, but it is fixed in my working branch.   I also spent some time this morning making sure that there were not any other Windows build issues and that things still run OK on Windows.  (They do)

I also normally check things on Windows before doing the (squashed) push; I just forgot this time.

mark-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1012 @@
+    // We're going to make a copy of at least one of the shader stage lists and change the pointer(s)
+    // to the copied stage lists in the pipeline create info structs.
+    // Make a (shallow) copy of the entire list of pipeline create info structs.
+    std::vector<VkGraphicsPipelineCreateInfo> copy_graphics_pipeline_infos(pCreateInfos, pCreateInfos + count);
```
Not sure how much it'll save/cost, but you also have available `safe_VkGraphicsPipelineCreateInfo`.

mark-lunarg commented on layers/core_validation.cpp:
```diff
@@ -5857,9 +5949,14 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateGraphicsPipelines(VkDevice device, VkPipeli
-    auto result =
-        dev_data->dispatch_table.CreateGraphicsPipelines(device, pipelineCache, count, pCreateInfos, pAllocator, pPipelines);
-
+    VkResult result;
```
GPUOverrideDispatchCreateGraphicsPipelines should return the modified parameters which will then be used in the call to CreateGraphicsPipelines instead of having two separate down-chain CGP calls. This would also allow you to keep the lock during your data munging (if needed), but make the down-chain CGP call while unlocked.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6717,11 +6827,17 @@ static void PostCallRecordCreatePipelineLayout(layer_data *dev_data, const VkPip
-    VkResult result = dev_data->dispatch_table.CreatePipelineLayout(device, pCreateInfo, pAllocator, pPipelineLayout);
+    if (GetEnables(dev_data)->gpu_validation) {
+        unique_lock_t lock(global_lock);
+        result = GpuCreatePipelineLayout(dev_data, pCreateInfo, pAllocator, pPipelineLayout);
```
Given that the supporting down-chain calls to date have been generally light weight (property/feature queries) that's reasonable.  With heavier weight calls (adding to command buffers/potentially allocating or free resources/fences, submit, wait, barrier, mapping and the like) probably not desirable to retain the layer lock.

Some of it is perf paranoia,  holding locks across heavyweight down-chain calls could have a big enough Heisenbug Effect to change results, potentially masking issues. (also a dose of lock ordering and long-jump fears)

karl-lunarg replied:
I took a stab at this problem in the fix I made for the comment about having the QueueWaitIdle in the core_validation code for QueueSubmit.   (This change is in the current squashed branch)

I moved the QWI to GpuPostCallQueueSubmit and added a reference parm for `global_lock` to GpuPostCallQueueSubmit.  The call to GpuPostCallQueueSubmit is and must be made with that lock in the unlocked state.  Then GpuPostCallQueueSubmit can lock and unlock as needed and would leave it unlocked on return because the unique_lock goes out of scope.

In this case, GpuPostCallQueueSubmit starts off by calling QWI without the lock.  It then gets the lock to analyze the debug buffers and generate the report.  The lock is unlocked on return.

(GpuPostCallQueueSubmit is fairly heavyweight because it does a mapping, pokes around a bit in the SPIR-V, and possibly does searches through shader source.  But if needed, I can pass the global_lock to `ProcessInstrumentationBuffer` to perhaps lock with finer granularity.)

Does this seem like a good pattern to follow for the rest of the functions that have this challenge?

jzulauf-lunarg replied:
for this PR -- fixing the down-chain dispatch for the API call, but not the support calls (except low-hanging fruit) would be sufficient, but we should review this, profiling/fixing any heavyweight cases.

karl-lunarg replied:
I'm resolving these lock issues by passing the global_lock as an extra param to the Gpu* functions and then dealing with locking and unlocking inside the Gpu* functions as needed.  The lock should be in the unlocked state when calling the Gpu* functions.  I've been told that this is more than enough to cover the disparate suggestions in this thread.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -5831,9 +5937,13 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateGraphicsPipelines(VkDevice device, VkPipeli
-
+    VkResult result;
+    if (GetEnables(dev_data)->gpu_validation) {
+        result = GpuCreateGraphicsPipelines(dev_data, pipelineCache, count, pCreateInfos, pAllocator, pPipelines);
```
Fixed by passing the lock object to the Gpu* routine and handling the locking in the Gpu routine.  This takes an extra parameter but is pretty clear.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+                                 const VkAllocationCallbacks *pAllocator, VkPipelineLayout *pPipelineLayout) {
+    auto gpu_state = GetGpuValidationState(dev_data);
+    if (gpu_state->aborted) {
+        return GetDispatchTable(dev_data)->CreatePipelineLayout(GetDevice(dev_data), pCreateInfo, pAllocator, pPipelineLayout);
```
Fixed by passing the lock object to the Gpu* routine and handling the locking in the Gpu routine. This takes an extra parameter but is pretty clear.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+        new_create_info.setLayoutCount = gpu_state->adjusted_max_desc_sets;
+    }
+    VkResult result;
+    result = GetDispatchTable(dev_data)->CreatePipelineLayout(GetDevice(dev_data), &new_create_info, pAllocator, pPipelineLayout);
```
Fixed by passing the lock object to the Gpu* routine and handling the locking in the Gpu routine. This takes an extra parameter but is pretty clear.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+    if (iter != cb_state->lastBound.end()) {
+        auto pipeline_state = iter->second.pipeline_state;
+        if (pipeline_state && (pipeline_state->pipeline_layout.set_layouts.size() <= gpu_state->desc_set_bind_index)) {
+            GetDispatchTable(dev_data)->CmdBindDescriptorSets(
```
Fixed by passing the lock object to the Gpu* routine and handling the locking in the Gpu routine. This takes an extra parameter but is pretty clear.

karl-lunarg commented on layers/gpu_validation.cpp:
```diff
@@ -0,0 +1,1118 @@
+   public:
+    GpuDescriptorSetManager(layer_data *dev_data) { this->dev_data_ = dev_data; }
+
+    ~GpuDescriptorSetManager() {
```
I went ahead and moved this out of the destructor and improved the locking in GpuPreCallRecordDestroyDevice to lock when updating state and unlocking when calling the driver.

In looking at other functions not mentioned in this review (like the Allocate/Free CB), these functions are running locked but still make driver calls.  These driver calls may not happen all the time and are pretty buried within several call levels.

I'm not that convinced that unlocking is that necessary when "calling down the chain" to another layer or the driver.  The lock is restricted in scope to just the core validation layer.  And the driver call is made without the rest of CV knowing about it.  There are no CV state updates related to these calls.  So calling down the chain while locked should be OK.

I just talked to @mark-lunarg and he says that these locking changes are going to make it hard to integrate with the upcoming chassis changes.  In addition, he agrees that we don't need to unlock to make driver calls.

So, I'm _not_ going to make this change to DestroyDevice and _not_ make the other changes I replied about this morning related to locking.

karl-lunarg commented:
I've made two additional pushes:

  karl-gpu-squashed            473aca2 [karl-lunarg/karl-gpu-squashed]
  karl-gpu-squashed1          4197bcb [karl-lunarg/karl-gpu-squashed1]
  karl-gpu-squashed2          b07e56e [karl-lunarg/karl-gpu-squashed2]
  karl-gpu-squashed3          5887289 [karl-lunarg/karl-gpu-squashed3]
  karl-gpu-squashed4          2b4934b [karl-lunarg/karl-gpu-squashed4]
  karl-gpu-squashed5          590a44b [karl-lunarg/karl-gpu-squashed5]
  karl-gpu-squashed6          473aca2 [karl-lunarg/karl-gpu-squashed6]

squashed5 addresses the comments that were not addressed in squashed4.

squashed6 is just a copyright update.

chrisforbes commented on docs/gpu_validation.md:
```diff
@@ -0,0 +1,707 @@
+| Stage  | Word 0           | Word 1     |
+|--------|------------------|------------|
+|Vertex  |VertexID          |InstanceID  |
+|Tess*   |InvocationID      |unused      |
```
Tess setup doesn't seem good. InvocationID is only available in the control shader, and is local within the patch. You want PrimitiveID to know which patch was involved. In the eval shader, you probably want the domain parameters to identify the domain point being tessellated. There's not enough space for them here though.

So this should probably be:
Tess Control: PrimitiveID InvocationID                       (yay, same as Geom)
Tess Eval: PrimitiveID <domain parameters>

greg-lunarg replied:
I will make the above tessellation changes to the instrumenter spirv-opt. The next time that the layers pick up a spirv-tools, GPU-assisted validation will make the changes.

How much space is needed to pass the domain parameters. Is it a fixed number? If not, is there a maximum?

greg-lunarg replied:
@chrisforbes One other change we think we need to make is for Vertex shaders to output what is known in SPIR-V as VertexIndex and InstanceIndex rather than VertexId and InstanceId because it appears that Vulkan1.1 does not support VertexIed and InstanceId. Does this seem right to you

greg-lunarg replied:
I would probably think that we don't want to validate instrumented shaders as a matter of course because if there was a problem it would not be a user problem so it wouldn't make sense for them to see it. I think our unit regression testing should be sufficient.

greg-lunarg replied:
> 2 parameters (u,v).

Cool! I will look into adding these.

chrisforbes replied:
Greg: this is PointCoord. There are actually 3 parameters for the triangle domain, but you can probably get away with not capturing w (should be normalized barycentric coords).
